File: sdk/node/native-runtime.js
'use strict'

const fs = require('node:fs')
const path = require('node:path')
const crypto = require('node:crypto')

const runtimeEnvNames = [
  'MESHLLM_NATIVE_RUNTIME_ARTIFACT_DIR',
  'MESHLLM_NATIVE_RUNTIME_DIR',
  'MESH_SDK_NATIVE_RUNTIME_DIR'
]

function resolveNativeRuntime(options = {}) {
  const candidates = []
  if (options.artifactDir) candidates.push(options.artifactDir)
  for (const name of runtimeEnvNames) {
    if (process.env[name]) candidates.push(process.env[name])
  }
  if (process.env.MESHLLM_NATIVE_RUNTIME_LIBRARY) {
    candidates.push(path.dirname(path.dirname(process.env.MESHLLM_NATIVE_RUNTIME_LIBRARY)))
  }
  for (const dir of options.searchDirs || []) candidates.push(dir)
  candidates.push(path.join(process.cwd(), 'meshllm-native'))
  candidates.push(path.join(process.cwd(), 'native'))

  const errors = []
  const seen = new Set()
  for (const candidate of candidates) {
    for (const artifactDir of artifactCandidates(candidate)) {
      const normalized = path.resolve(artifactDir)
      if (seen.has(normalized)) continue
      seen.add(normalized)
      try {
        return validateNativeRuntime(normalized)
      } catch (error) {
        errors.push(`${normalized}: ${error.message}`)
      }
    }
  }

  const detail = errors.length === 0
    ? 'no candidate native runtime artifact directories were configured'
    : errors.join('; ')
  throw new Error(`MeshLLM native runtime artifact not found: ${detail}`)
}

function validateNativeRuntime(artifactDir) {
  const normalized = path.resolve(artifactDir)
  const manifestPath = path.join(normalized, 'manifest.json')
  if (!fs.existsSync(normalized) || !fs.statSync(normalized).isDirectory()) {
    throw new Error('artifact directory does not exist')
  }
  if (!fs.existsSync(manifestPath)) {
    throw new Error('manifest.json does not exist')
  }

  const manifest = JSON.parse(fs.readFileSync(manifestPath, 'utf8'))
  const library = path.resolve(normalized, requiredString(manifest, 'library'))
  const expected = requiredString(manifest, 'library_sha256').toLowerCase()
  if (!fs.existsSync(library)) {
    throw new Error(`native library does not exist: ${library}`)
  }
  const actual = sha256(library)
  if (actual !== expected) {
    throw new Error(`native library checksum mismatch for ${library}`)
  }

  return {
    artifactId: requiredString(manifest, 'artifact_id'),
    artifactDir: normalized,
    manifest: manifestPath,
    library,
    metadata: manifest
  }
}

function artifactCandidates(candidate) {
  if (!candidate) return []
  const normalized = path.resolve(candidate)
  if (fs.existsSync(path.join(normalized, 'manifest.json'))) {
    return [normalized]
  }
  if (!fs.existsSync(normalized) || !fs.statSync(normalized).isDirectory()) {
    return [normalized]
  }
  const children = fs.readdirSync(normalized)
    .map((entry) => path.join(normalized, entry))
    .filter((entry) => fs.statSync(entry).isDirectory() && path.basename(entry).startsWith('meshllm-native-'))
    .sort()
  return [normalized, ...children]
}

function requiredString(object, key) {
  if (typeof object[key] !== 'string' || object[key].length === 0) {
    throw new Error(`manifest field missing or not a string: ${key}`)
  }
  return object[key]
}

function sha256(file) {
  return crypto.createHash('sha256').update(fs.readFileSync(file)).digest('hex')
}

module.exports = {
  resolveNativeRuntime,
  validateNativeRuntime
}

File: sdk/kotlin/src/main/kotlin/ai/meshllm/NativeRuntime.kt
package ai.meshllm

import java.io.File
import java.security.MessageDigest

data class NativeRuntimeConfig(
    val artifactDir: File? = null,
    val searchDirs: List<File> = emptyList(),
)

data class NativeRuntimeArtifact(
    val artifactId: String,
    val artifactDir: File,
    val manifest: File,
    val library: File,
    val uniffiLibrary: File,
)

object NativeRuntime {
    private const val COMPONENT_LIBRARY_OVERRIDE = "uniffi.component.mesh_ffi.libraryOverride"

    fun configure(config: NativeRuntimeConfig = NativeRuntimeConfig()): NativeRuntimeArtifact {
        val artifact = resolve(config)
        System.setProperty(COMPONENT_LIBRARY_OVERRIDE, artifact.uniffiLibrary.absolutePath)
        return artifact
    }

    fun resolve(config: NativeRuntimeConfig = NativeRuntimeConfig()): NativeRuntimeArtifact {
        val candidates = buildList {
            config.artifactDir?.let(::add)
            systemPath("meshllm.nativeRuntime.artifactDir")?.let(::add)
            systemPath("meshllm.nativeRuntime.dir")?.let(::add)
            envPath("MESHLLM_NATIVE_RUNTIME_ARTIFACT_DIR")?.let(::add)
            envPath("MESHLLM_NATIVE_RUNTIME_DIR")?.let(::add)
            envPath("MESH_SDK_NATIVE_RUNTIME_DIR")?.let(::add)
            envPath("MESHLLM_NATIVE_RUNTIME_LIBRARY")?.parentFile?.parentFile?.let(::add)
            addAll(config.searchDirs)
            add(File(System.getProperty("user.dir"), "meshllm-native"))
            add(File(System.getProperty("user.dir"), "native"))
        }

        val errors = mutableListOf<String>()
        for (candidate in candidates.distinctBy { it.normalized().path }) {
            for (artifactDir in artifactCandidates(candidate)) {
                val result = runCatching { validate(artifactDir) }
                result.getOrNull()?.let { return it }
                result.exceptionOrNull()?.message?.let { errors.add("${artifactDir.path}: $it") }
            }
        }

        val detail = if (errors.isEmpty()) {
            "no candidate runtime artifact directories were configured"
        } else {
            errors.joinToString("; ")
        }
        throw IllegalStateException("MeshLLM native runtime artifact not found: $detail")
    }

    fun validate(artifactDir: File): NativeRuntimeArtifact {
        val normalizedDir = artifactDir.normalized()
        val manifest = normalizedDir.resolve("manifest.json")
        require(normalizedDir.isDirectory) { "artifact directory does not exist" }
        require(manifest.isFile) { "manifest.json does not exist" }

        val manifestText = manifest.readText()
        val artifactId = stringField(manifestText, "artifact_id")
        val libraryRelativePath = stringField(manifestText, "library")
        val uniffiLibraryRelativePath = stringField(manifestText, "uniffi_library")
        val expectedSha256 = stringField(manifestText, "library_sha256").lowercase()

        val library = normalizedDir.resolve(libraryRelativePath).normalized()
        val uniffiLibrary = normalizedDir.resolve(uniffiLibraryRelativePath).normalized()
        require(library.isFile) { "native library does not exist: ${library.path}" }
        require(uniffiLibrary.isFile) { "UniFFI library does not exist: ${uniffiLibrary.path}" }

        val actualLibrarySha256 = sha256(library)
        require(actualLibrarySha256 == expectedSha256) {
            "native library checksum mismatch for ${library.path}"
        }
        require(sha256(uniffiLibrary) == expectedSha256) {
            "UniFFI library checksum mismatch for ${uniffiLibrary.path}"
        }

        return NativeRuntimeArtifact(
            artifactId = artifactId,
            artifactDir = normalizedDir,
            manifest = manifest,
            library = library,
            uniffiLibrary = uniffiLibrary,
        )
    }

    private fun systemPath(name: String): File? =
        System.getProperty(name)?.takeIf(String::isNotBlank)?.let(::File)

    private fun envPath(name: String): File? =
        System.getenv(name)?.takeIf(String::isNotBlank)?.let(::File)

    private fun artifactCandidates(candidate: File): List<File> {
        val normalized = candidate.normalized()
        if (normalized.resolve("manifest.json").isFile) {
            return listOf(normalized)
        }
        val children = normalized.listFiles()
            ?.filter { it.isDirectory && it.name.startsWith("meshllm-native-") }
            ?.sortedBy(File::getName)
            .orEmpty()
        return listOf(normalized) + children
    }

    private fun stringField(json: String, key: String): String {
        val pattern = Regex("\"${Regex.escape(key)}\"\\s*:\\s*\"((?:\\\\.|[^\"])*)\"")
        val match = pattern.find(json) ?: error("manifest field missing or not a string: $key")
        return match.groupValues[1]
            .replace("\\/", "/")
            .replace("\\\"", "\"")
            .replace("\\\\", "\\")
    }

    private fun sha256(file: File): String {
        val digest = MessageDigest.getInstance("SHA-256")
        file.inputStream().use { input ->
            val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
            while (true) {
                val read = input.read(buffer)
                if (read < 0) {
                    break
                }
                digest.update(buffer, 0, read)
            }
        }
        return digest.digest().joinToString("") { "%02x".format(it) }
    }

    private fun File.normalized(): File = absoluteFile.toPath().normalize().toFile()
}
